src/coreclr/src/jit/optimizer.cpp

```diff
@@ -9189,3 +9189,132 @@ void Compiler::optOptimizeBools()
     fgDebugCheckBBlist();
 #endif
 }
+
+typedef JitHashTable<unsigned, JitSmallPrimitiveKeyFuncs<unsigned>, unsigned> LclVarRefCounts;
+
+//------------------------------------------------------------------------------------------
+// optRemoveRedundantZeroInits: Remove redundant zero intializations.
+//
+// Notes:
+//    This phase iterates over basic blocks starting with the first basic block until there is no unique
+//    basic block successor or until it detects a loop. It keeps track of local nodes it encounters.
+//    When it gets to an assignment to a local variable or a local field, it checks whether the assignment
+//    is the first reference to the local (or to the parent of the local field), and, if so,
+//    it may do one of two optimizations:
+//      1. If the local is untracked, the rhs of the assignment is 0, and the local is guaranteed
+//         to be fully initialized in the prolog, the explicit zero initialization is removed.
+//      2. If the assignment is to a local (and not a field) and either the local has no gc pointers or there
+//         are no gc-safe points between the prolog and the assignment, it marks the local with lvHasExplicitInit
+//         which tells the codegen not to insert zero initialization for this local in the prolog.
+
+void Compiler::optRemoveRedundantZeroInits()
+{
+#ifdef DEBUG
+    if (verbose)
+    {
+        printf("*************** In optRemoveRedundantZeroInits()\n");
+    }
+#endif // DEBUG
+
+    CompAllocator   allocator(getAllocator(CMK_ZeroInit));
+    LclVarRefCounts refCounts(allocator);
+    bool            hasGCSafePoint = false;
+
+    assert(fgStmtListThreaded);
+
+    for (BasicBlock* block = fgFirstBB; (block != nullptr) && ((block->bbFlags & BBF_MARKED) == 0);
+         block             = block->GetUniqueSucc())
+    {
+        block->bbFlags |= BBF_MARKED;
+        for (Statement* stmt = block->FirstNonPhiDef(); stmt != nullptr;)
+        {
+            Statement* next = stmt->GetNextStmt();
+            for (GenTree* tree = stmt->GetTreeList(); tree != nullptr; tree = tree->gtNext)
+            {
+                if ((tree->gtFlags & GTF_CALL) != 0)
+                {
+                    hasGCSafePoint = true;
+                }
+
+                switch (tree->gtOper)
+                {
+                    case GT_LCL_VAR:
+                    case GT_LCL_FLD:
+                    case GT_LCL_VAR_ADDR:
+                    case GT_LCL_FLD_ADDR:
+                    {
```

Comment:
For non-address exposed locals we only care about seeing defs, not uses?

Reply:
We care about seeing defs when removing explicit zero initializations and we care about seeing uses when marking locals with `lvHasExplicitInit`. I tried a change where I tracked defs and uses separately and saw no new diffs in framework or benchmarks so will keep it simple.